analytical_engine/apps/voterank/voterank.h

```diff
@@ -0,0 +1,224 @@
+
+    // assign initial ranks and weights
+    std::vector<long unsigned int> edge_nums(thread_num(), 0);
+    ForEach(inner_vertices, [&ctx, &edge_nums,&frag, &messages](int tid, vertex_t u) {
+      int EdgeNum = frag.GetOutgoingAdjList(u).Size();
+      edge_nums[tid] += EdgeNum;
+      ctx.rank[u] = 0;
+      ctx.weight[u] = 1.0;
+      ctx.scores[u] = 0.0;
+      messages.SendMsgThroughIEdges<fragment_t, double>(frag, u,
+                                                          ctx.weight[u], tid); 
+    });
+    long unsigned int sumEdgeNum = 0;
+    for(auto i : edge_nums){
+      sumEdgeNum += i;
+    }
+    Sum(sumEdgeNum,sumEdgeNum);
+    ctx.avg_degree = static_cast<double>(sumEdgeNum) /
+                 static_cast<double>(graph_vnum);
+    
+#ifdef PROFILING
+    ctx.exec_time += GetCurrentTime();
+    ctx.postprocess_time -= GetCurrentTime();
+#endif
+
+#ifdef PROFILING
+    ctx.postprocess_time += GetCurrentTime();
+#endif
+    messages.ForceContinue();
+  }
+
+
+  void IncEval(const fragment_t& frag, context_t& ctx,
+               message_manager_t& messages) {
+    auto inner_vertices = frag.InnerVertices();
+    ++ctx.step;
+    if (ctx.step > ctx.max_round) {
+      return;
+    }
+    
+
+#ifdef PROFILING
+    ctx.exec_time -= GetCurrentTime();
+#endif
+
+    // pull weights from inner neighbors
+    vertex_t v;
+    
+    ForEach(inner_vertices, [&ctx, &frag](int tid, vertex_t u) {
+      if (ctx.rank[u] == 0) {
+        double cur = 0;
+        auto es = frag.GetOutgoingInnerVertexAdjList(u);
+        for (auto& e : es) {
+          cur += ctx.weight[e.get_neighbor()];
+        }
+        ctx.scores[u] = cur;
+      }
+    });
+
+#ifdef PROFILING
+    ctx.exec_time += GetCurrentTime();
+    ctx.preprocess_time -= GetCurrentTime();
+#endif
+
+    // process received weights sent by other workers
+    {
+      messages.ParallelProcess<fragment_t, double>(
+          thread_num(), frag, [&ctx](int tid, vertex_t u, const double& msg) {
+            ctx.weight[u] = msg;
+          });
+    }
+    
+#ifdef PROFILING
+    ctx.preprocess_time += GetCurrentTime();
+    ctx.exec_time -= GetCurrentTime();
+#endif
+
+    
+    auto compare = [](std::pair<double,oid_t>& lhs, 
+                 const std::pair<double,oid_t>& rhs) { 
+              const double EPS = 1e-8;
+              if(fabs(lhs.first - rhs.first) < EPS){
+                 if(rhs.second < lhs.second){
+                    lhs = rhs;
+                  }
+              } else if(lhs.first < rhs.first){
+                  lhs = rhs;
+              }
+          };
+    // compute new scores
+    std::vector<std::pair<double,oid_t>> max_scores(thread_num(),{0,{}});
+    ForEach(inner_vertices, [&ctx,compare, &max_scores,&frag](int tid, vertex_t u) {
+        if (ctx.rank[u] == 0) {
+          double cur = ctx.scores[u];
+          auto es = frag.GetOutgoingOuterVertexAdjList(u);
+          for (auto& e : es) {
+            cur += ctx.weight[e.get_neighbor()];
```

Comment:
would that better if  we compute the weight sum one time and keep in a context member ?  It could avoid to recompute the weight sum of u